websites/management/commands/rename_by_text_id.py

```diff
@@ -0,0 +1,55 @@
+from pathlib import Path
+
+import boto3
+from django.conf import settings
+from django.core.management import BaseCommand
+from django.utils.text import slugify
+
+from gdrive_sync.models import DriveFile
+from main.utils import get_dirpath_and_filename
+from websites.models import WebsiteContent
+
+
+class Command(BaseCommand):
+    """Rename the file on S3 associated with the WebsiteContent object to a new filename, and update the object and corresponding DriveFile."""
+
+    help = __doc__
+
+    def add_arguments(self, parser):
+
+        parser.add_argument(
+            "--text_id",
+            dest="text_id",
+            help="text_id of WebsiteContent object to be updated",
+            required=True,
+        )
+
+        parser.add_argument(
+            "--new_filename",
+            dest="new_filename",
+            help="New filename to be associated with the WebsiteContent object",
+            required=True,
+        )
+
+    def handle(self, *args, **options):
+        obj = WebsiteContent.objects.get(text_id=options["text_id"])
+        df = DriveFile.objects.get(resource=obj)
+        s3 = boto3.resource("s3")
+        # slugify just the provided name and then make the extensions lowercase
+        filepath = Path(options["new_filename"])
+        basename = options["new_filename"].rstrip("".join(filepath.suffixes))
+        new_filename = slugify(basename)
+        if filepath.suffixes:
+            new_filename += "".join(filepath.suffixes).lower()
+        df_path = df.s3_key.split("/")
+        df_path[-1] = new_filename
+        new_key = "/".join(df_path)
+        s3.Object(settings.AWS_STORAGE_BUCKET_NAME, new_key).copy_from(
+            CopySource=settings.AWS_STORAGE_BUCKET_NAME + "/" + df.s3_key
+        )
+        s3.Object(settings.AWS_STORAGE_BUCKET_NAME, df.s3_key).delete()
+        df.s3_key = new_key
+        obj.file = new_key
+        obj.filename = get_dirpath_and_filename(new_filename)[1]
```

Comment:
I got an integrity error when running this a certain way, which I think might match the scenario under which the problem it's trying to fix happened in production:

- Upload a file to google drive ("word_sample_1.doc")
- Sync, creating `Gdrive` and `WebsiteContent` objects.
- Delete the file from google drive, then upload it again.
- Sync again.  Now I have two resources, with filenames "word_sample_1.doc" (original) and "word_sample_12.doc" (reuploaded)

I created links to both resources in a page.  I also added links to both resources in the site menu.

Run `docker-compose run web python manage.py rename_by_text_id --text_id  <text_id of reuploaded content> --new_filename word_sample_1.doc`

Because the original resource still exists, and has a `filename` of `word_sample_1` which must be unique within each site, I get an `IntegrityError`.

```
django.db.utils.IntegrityError: duplicate key value violates unique constraint "unique_page_content_destination"
DETAIL:  Key (website_id, dirpath, filename)=(538d4c8b-3914-4685-95b0-b5fa48db9c5f, content/resources, word_sample_1) already exists.
```

So either the rename can't be the same as that of original (which I think might defeat the purpose?) or the old `WebsiteContent` object also needs a rename, or the old one needs to be deleted, which requires checking for links/embeds and changing those to the new resource instead.  Might be good to get some feedback from Peter on what's the most ideal here from the user perspective.

I tried again using a slightly different rename (`word_sample__1`) just to proceed further, republished, and noticed another issue with site menu.  [The menu is still pointing to the old url](https://github.com/mbtestorg2022/course9/blob/main/config/_default/menus.yaml#L12).  However, the link within the page is still pointing to the new, correct url.  So I think there might need to be a check added to see if the resource is linked to from the site menu, and if so update it.

Reply:
Possible approach to this:
- check if a WebsiteContent object with the `new_filename` exists in that course.
- if it does not, proceed as currently in the PR.
- if it does, check whether it exists in any course content (links or embedded content).
    - if it exists in any course content, print the name of the course content and abort
    - else: print a warning that the old object is being deleted and then proceed as currently in the PR.

Reply:
Note that there should be at most 1 existing `WebsiteContent` object with this filename already in the database, because otherwise there would have already been an `IntegrityError`, as above.

Reply:
Thanks Paul. I think deleting the resource fits the use case. Do we have a method for deleting resources? I've gotten requests for that as well, but I haven't been prioritizing them.

Reply:
We can directly delete the `WebsiteContent` object `obj` with `obj.delete()`. The file on S3 would not need to be deleted, since it's being written over by the renaming command.

Also, just documenting that we are not going to currently support issues with broken links due to a file rename either in navigation menus or in pages, because:

- Menus should be linking to pages, not resources.
- Linking to an internal URL in a page instead of using a resource link is discouraged.